### PR TITLE
[quant] per channel quantization support

### PR DESCRIPTION
**Pull request by dskhudia**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#24936 [quant] per channel quantization support**
* #24935 [quant] Per Channel quantization APIs
* #24934 [quant] Add missing functions and methods for channelwise quantization

Per channel quantization support in qconv2d + tests

Differential Revision: [D16909378](https://our.internmc.facebook.com/intern/diff/D16909378/)

### CODE REVIEW

File: aten/src/ATen/native/quantized/cpu/qconv_unpack.cpp [hunk 1]
```diff
@@ -33,13 +33,34 @@ class QConvUnpackWeightsInt8 final : public c10::OperatorKernel {
 
     int C_per_G = input_channels / groups;
 
+
```
Comment (jianyuh) [inline aten/src/ATen/native/quantized/cpu/qconv_unpack.cpp]: nit: extra line.

File: test/test_quantized.py [hunk 2]
```diff
@@ -784,12 +783,13 @@ class TestQuantizedConv(unittest.TestCase):
            dilation=st.integers(1, 1),
            X_scale=st.floats(0.2, 1.6),
            X_zero_point=st.integers(0, 4),
-           W_scale=st.floats(0.2, 1.6),
-           W_zero_point=st.integers(-5, 5),
+           W_scale=st.lists(st.floats(0.2, 1.6), min_size=1, max_size=2),
+           W_zero_point=st.lists(st.integers(-5, 5), min_size=1, max_size=2),
```
[the remark this answers is not in the document]
Comment (dskhudia) [reply test/test_quantized.py]: It may. Those cases are handled.

**Review by jianyuh — APPROVED** [review]
Looks good!

File: aten/src/ATen/native/quantized/cpu/fbgemm_utils.h [hunk 3]
```diff
@@ -1,5 +1,7 @@
 #pragma once
 
+#include <c10/core/QScheme.h>
+
```
Comment (raghuramank100) [inline aten/src/ATen/native/quantized/cpu/fbgemm_utils.h]: NIT: blank line

File: aten/src/ATen/native/quantized/cpu/qconv_prepack.cpp [hunk 4]
```diff
@@ -59,7 +59,17 @@ class QConvPackWeightInt8 final : public c10::OperatorKernel {
...
     auto weight_contig = weight.contiguous();
-    int32_t weight_zero_point_int32 = weight.q_zero_point();
+    const auto qtype = weight.qscheme();
+    std::vector<int32_t> zero_points(1, 0);
+    if (qtype == kPerTensorAffine) {
+      zero_points[0] = weight.q_zero_point();
+    } else if (qtype == kPerChannelAffine) {
+      zero_points.resize(output_channels, 0);
+      for (int i = 0; i < output_channels; ++i) {
+        zero_points[i] = weight.q_per_channel_zero_points()[i].item<int32_t>();
```
Comment (raghuramank100) [inline aten/src/ATen/native/quantized/cpu/qconv_prepack.cpp]: Is there a more efficient way to do this?  cc @jerryzh168

File: test/test_quantized.py [hunk 5]
```diff
@@ -889,8 +903,20 @@ def test_qconv(
...
 
         X_q = torch.quantize_linear(X_NHWC, scale=X_scale, zero_point=X_zero_point, dtype=torch.quint8)
-        W_q = torch.quantize_linear(W_KRSC, scale=W_scale, zero_point=W_zero_point, dtype=torch.qint8)
-        b_q = torch.quantize_linear(b, scale=X_scale * W_scale, zero_point=0, dtype=torch.qint32) if use_bias else None
+        if use_channelwise:
+            W_q = torch.quantize_linear_per_channel(W_KRSC,
+                                                    W_scales_tensor.to(dtype=torch.double),
+                                                    W_zero_points_tensor.to(dtype=torch.long),
+                                                    [0],
+                                                    dtype=torch.qint8)
```
Comment (raghuramank100) [inline test/test_quantized.py]: Can we sweep over qint8 and quint8 for W?

**Review by raghuramank100 — APPROVED** [review]
Please add extra tests.

**Comment by zrphercule** [comment]
Will revert this one since it is breaking master.

**Pull request closed without merging**